src/librustc_typeck/collect.rs

```diff
@@ -1978,10 +1980,19 @@ fn explicit_predicates_of<'tcx>(
                         SizedByDefault::Yes,
                         tcx.def_span(def_id),
                     );
-
                     let bounds_predicates = bounds.predicates(tcx, opaque_ty);
+
                     if impl_trait_fn.is_some() {
-                        // opaque types
+                        // An `impl Trait` type
+                        let bounds_predicates = tcx.infer_ctxt().enter(|infcx| {
+                            fully_normalize(
+                                &infcx,
+                                FulfillmentContext::new(),
+                                ObligationCause::dummy(),
+                                tcx.param_env(def_id),
+                                &bounds_predicates,
+                            ).expect("failed to normalize opaque type bounds")
+                        });
```

Comment:
Hmm, I don't think this is the right place to do normalization.

Reply:
In particular, the result of `explicit_predicates_of` is unnormalized; we normally do normalization after substitution has been performed.

Reply:
The question then is where is the *right* place

Reply:
I initially tried doing this somewhere in `rustc_typeck::check::check_fn`, but that only lead to ICEs later on (during MIR type checking I believe). If that's a better place I can do that again and try fixing the ICEs.

Reply:
It might be -- perhaps push that version so I can see exactly where you mean?

Reply:
See [my comment here](https://github.com/rust-lang/rust/issues/60414#issuecomment-508044135) =)